### DISCUSSION TITLE
Respect GNOME Large Text setting (org.gnome.desktop.interface text-scaling-factor)

### DISCUSSION THREAD

**Discussion by hpfr** [ghostty-org/ghostty#3771] [Feature Requests, Ideas] [answered]
In GNOME Settings, under Accessibility/Seeing, there is a Large Text toggle. Under the hood, this sets `org.gnome.desktop.interface text-scaling-factor` to `1.25`. GTK respects this, so the title bar and context menus look correct, but the terminal emulator text itself does not.

VTE apparently respects this [via Pango](https://gitlab.gnome.org/GNOME/vte/-/commit/bfd656e49e9151207d7291d60dfefacff52e73f0). For a “native” experience on GNOME, Ghostty probably should too. If Pango isn’t used, reading the key via [GSettings](https://docs.gtk.org/gio/class.Settings.html) might be the way to go.

**Comment by hpfr** [comment]
This did come up as a follow-up to the PR which introduced support for interface-wide scaling, see https://github.com/ghostty-org/ghostty/pull/746#issuecomment-1783868921.

Mitchell seems amenable to support in https://github.com/ghostty-org/ghostty/pull/746#issuecomment-1783881539. Without Pango, I believe the GSettings key mentioned above would be the GTK way to figure out the user’s Large Text setting. Note that while GNOME Settings exposes a simple toggle that changes the value from 1 to 1.25, users can set other values via the `gsettings` CLI or other interfaces, so we should use the actual value that is read.

[Gio.Settings::changed](https://docs.gtk.org/gio/signal.Settings.changed.html) may allow reacting to changes to the key while running (as native GTK text does)

**Reply by hpfr** [reply]
It looks like GTK/GSettings alone does not guarantee write persistence or the availability of desktop schemas like `org.gnome.desktop.interface`, the default backend is `memory` and there is no hard dependency on `dconf` or the desktop schemas.

https://github.com/NixOS/nixpkgs/blob/394571358ce82dff7411395829aa6a3aad45b907/doc/languages-frameworks/gnome.section.md?plain=1#L11

Looks like Arch does make dconf and the schemas a GTK dependency according to [GNOME package guidelines](https://wiki.archlinux.org/title/GNOME_package_guidelines#GSettings_schemas), but based on the Nix docs, it’s not a hard upstream dependency so we may not be able to rely on it without declaring more dependencies for packagers.

Anyway, we’re not writing here, so this all probably just means we need to make sure we fail gracefully if the schema isn’t available before calling [`g_settings_new`](https://docs.gtk.org/gio/ctor.Settings.new.html), as documented there.

**Comment by tristan957** [comment]
This would be easy to setup. Since Pango is a dependency of GTK, we should just use Pango as well. No sense in re-inventing the wheel.

**Reply by tristan957** [reply]
Here are the Pango docs: https://docs.gtk.org/Pango. It would be useful if you could link to the appropriate APIs we would need to call.

**Reply by hpfr** [reply]
I’m not a GTK expert, but I think Pango is used (largely internally) by GTK and libadwaita for text rendering in native widgets (which is why the GTK decorations already respect the large text setting). I don’t think using Pango is straightforward for the actual terminal emulator text, since libghostty rolls its own font rendering and from GTK’s perspective it’s a mostly “opaque” GLArea, but correct me if I’m wrong. That’s why I think we need to read the GSettings key (and use it to modify `font-size` when initializing the Surface), since unlike typical GTK applications, we are rendering text outside of native widgets.

Perhaps a Pango rendering `Backend` can be implemented one day, but that seems like a substantial effort that probably shouldn’t block respecting this pref. Also, it looks like Pango doesn’t actually know anything about `text-scaling-factor` and is fed text sizes from the GTK widget machinery anyway, so I shouldn’t have brought it up (VTE is a GTK widget).

According to [widget: notify of style change upon css_changed (bfd656e4) · Commits · GNOME / vte · GitLab](https://gitlab.gnome.org/GNOME/vte/-/commit/bfd656e49e9151207d7291d60dfefacff52e73f0), [Gtk.Settings:gtk-xft-dpi](https://docs.gtk.org/gtk4/property.Settings.gtk-xft-dpi.html) is where GTK takes the scaling factor into account, so we can potentially read that and divide by `(96 * 1024)` ([96](https://gitlab.gnome.org/GNOME/gtk/-/blob/818d5952eed66136f9797b8ccb20ca2d44c907b1/gdk/wayland/gdkdisplay-wayland.c?page=2#L1564), [1024](https://gitlab.gnome.org/GNOME/gtk/-/blob/818d5952eed66136f9797b8ccb20ca2d44c907b1/gdk/wayland/gdkdisplay-wayland.c?page=2#L1679)), but I am not sure depending on some hardcoded GTK internals is better than just reading the key ourselves from GSettings if available. I might be able to look into drafting a patch for this soon.

**Comment by mitchellh** [comment, marked as answer]
Approved #4338
